sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RetryUtils.java

```diff
@@ -82,11 +84,20 @@
 
                 if (!shouldRetryResult.shouldRetry) {
                     retryPolicy.updateEndTime();
-                    if(shouldRetryResult.exception == null) {
-                        return Mono.error(e);
-                    } else {
-                        return Mono.error(shouldRetryResult.exception);
+
+                    final Throwable errorToReturn = shouldRetryResult.exception != null ? shouldRetryResult.exception : e;
+                    final Mono<T> failure = Mono.error(errorToReturn);
+
+                    if (shouldRetryResult.policyArg != null &&
+                        shouldRetryResult.policyArg.getValue0() != null &&
+                        shouldRetryResult.policyArg.getValue0()) {
+
+                        return addressSelector.resolveAddressesAsync(
```

Comment:
Why we need to refresh address here? If we reach here, it means we will not going to retry the request, is that we are trying to proactively refresh the address?

Reply:
It is the Gone for write requests - before we were refreshing addresses and retry. Retrying is a problem because writes aren't idempotent. But the address refresh as reaction of all 410s should still happen

Reply:
I see, got it, make sense, thanks for the explanation.